### PR TITLE
bugfix/11635-wheelscroll-click-tooltip

### LINKED ISSUE #11635: Scrolling doesn't trigger tooltip hide when cursor leaves the chart's container
#### Expected behaviour
1. consider there are 10 charts in one page from top to bottom.
2. Now scroll the page multiples time top to bottom using mouse wheel
3. All the tool-tips get clustered and do not disappear.

#### Actual behaviour
1. Tool tips should disappear and do not get clustered

#### Product version
    "highcharts": "7.1.2",

#### Affected browser(s)
Chrome

### PR DESCRIPTION
**Pull request by bre1470**
Fixed #11635, scrolling didn't trigger hiding of the tooltip when the cursor left the chart's container.

___

- Close #11635, Tooltip hides with wheelscroll click. https://jsfiddle.net/g643wuLj/
- Improved TestController.

### CODE REVIEW

File: ts/parts/Interaction.ts [hunk 1]
```diff
@@ -291,10 +291,20 @@ TrackerMixin = H.TrackerMixin = {
...
+            i: number,
+            onMouseOver = function (e: Highcharts.PointerEventObject): void {
+
+                pointer.normalize(e);
+
+                if (
+                    chart.hoverSeries !== series &&
+                    (
+                        !tooltip ||
+                        !tooltip.isStickyOnContact()
```
Comment (TorsteinHonsi) [inline ts/parts/Interaction.ts]: Is this related to #11635, or is it a random issue related to the `stickOnContact` feature? Ideally, this PR should be focused on only the changes required to fix #11635. We can leave it for now, I just want to understand the connection.

File: ts/parts/Pointer.ts [hunk 2]
```diff
@@ -1065,15 +1064,20 @@ class Pointer {
...
         this.isDirectTouch = false;
 
         if (
             series &&
             relatedTarget &&
             !series.stickyTracking &&
+            (
+                !tooltip ||
+                !tooltip.isStickyOnContact()
+            ) &&
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: As above, are we mixing in some unrelated fix?

File: ts/parts/Pointer.ts [hunk 3]
```diff
@@ -1283,15 +1287,21 @@ class Pointer {
         // Normalize before the 'if' for the legacy IE (#7850)
         e = this.normalize(e);
 
-        if ((e as any).button !== 2) {
+        this.onContainerMouseMove(e);
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: It seems a little strange to call `onContainerMouseMove()` from the `onContainerMouseDown` function... Why are we doing that? Do we know the side effects? Are there parts of the `onContainerMouseMove` function that we can call directly instead?

At the least, we should have a comment explaining why we do this strange call.

File: ts/parts/Pointer.ts [hunk 4]
```diff
@@ -1283,15 +1287,21 @@ class Pointer {
...
 
-            this.zoomOption(e);
+        if (
+            typeof e.button === 'undefined' ||
+            e.button === 0 // #11635, limiting to primary button
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: This logic now changed from `e.button !== 2` to `e.button === 0`. If I remember correctly, the primary button in IE is or was 1, hence the !== 2 part. Has this been tested with IE? If so, we should have a comment.

File: ts/parts/Pointer.ts [hunk 5]
```diff
@@ -1283,15 +1287,21 @@ class Pointer {
...
-            if (e.preventDefault) {
+            this.zoomOption(e);
+            /*
+            // #295, dragging not always working in Firefox
+            // #11635, turn off condition because of wheel click issue
+            if (H.isFirefox) {
                 e.preventDefault();
             }
-
+             */
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: Instead of commenting out obsolete code, we should remove it straight away. If a regression happens, we will find the cause using `git blame`.

File: ts/parts/Pointer.ts [hunk 6]
```diff
@@ -1333,13 +1351,16 @@ class Pointer {
...
 
+        /*
         if (
             !defined(H.hoverChartIndex) ||
             !charts[H.hoverChartIndex as any] ||
             !(charts[H.hoverChartIndex as any] as any).mouseIsDown
         ) {
             H.hoverChartIndex = chart.index;
         }
+         */
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: As above, obsolete code can safely be deleted rather than commented.

File: ts/parts/Pointer.ts [hunk 7]
```diff
@@ -2106,51 +2127,83 @@ class Pointer {
             ownerDoc = container.ownerDocument;
 
         container.onmousedown = function (e: MouseEvent): void {
-            pointer.onContainerMouseDown(e as any);
+            pointer.onContainerMouseDown(pointer.normalize(e));
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: `Pointer.onContainerMouseDown` already does `normalize` internally...

File: ts/parts/Pointer.ts [hunk 8]
```diff
@@ -2106,51 +2127,83 @@ class Pointer {
...
         };
         container.onmousemove = function (e: MouseEvent): void {
-            pointer.onContainerMouseMove(e as any);
+            pointer.onContainerMouseMove(pointer.normalize(e));
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: `Pointer.onContainerMouseMove` already does `normalize` internally...

File: ts/parts/Pointer.ts [hunk 9]
```diff
@@ -2106,51 +2127,83 @@ class Pointer {
...
         };
         container.onclick = function (e: MouseEvent): void {
-            pointer.onContainerClick(e as any);
+            pointer.onContainerClick(pointer.normalize(e));
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: `Pointer.onContainerClick` already does `normalize` internally...

File: ts/parts/Pointer.ts [hunk 10]
```diff
@@ -2106,51 +2127,83 @@ class Pointer {
...
                 }
             );
             if (!H.unbindDocumentTouchEnd) {
                 H.unbindDocumentTouchEnd = addEvent(
                     ownerDoc,
                     'touchend',
-                    pointer.onDocumentTouchEnd as any
+                    function (e: TouchEvent): void {
+                        pointer.onDocumentTouchEnd(pointer.normalize(e));
+                    }
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: With all the `addEvent` and `on[XXX]` assignments above, would it be better to just use the `Function.bind` function to bind it to the pointer?

Also, as noted, many of the already normalize the event arguments, should we do that with the rest where it is needed, so we can have simple assignments of bound functions?

File: ts/parts/Pointer.ts [hunk 11]
```diff
@@ -2106,51 +2127,83 @@ class Pointer {
...
+    /**
+     * Sets the index of the hovered chart and leaves the previous hovered
+     * chart, to reset states like tooltip.
+     *
+     * @private
+     * @function Highcharts.Pointer#setHoverChartIndex
+     */
+    public setHoverChartIndex(): void {
+        const chart = this.chart;
+        const hoverChart = H.charts[H.hoverChartIndex || -1];
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: Should be:
`const hoverChart = H.charts[pick(H.hoverChartIndex, -1)];`

... to prevent the 0 case failure.

File: ts/parts/Pointer.ts [hunk 12]
```diff
@@ -2106,51 +2127,83 @@ class Pointer {
...
+
+        if (
+            hoverChart &&
+            hoverChart.index !== chart.index
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: I think we can safely do `if (hoverChart && hoverChart !== chart)` ?

File: ts/parts/Pointer.ts [hunk 13]
```diff
@@ -2169,10 +2222,13 @@ class Pointer {
             pinchDown,
             isInside;
 
+        /*
         if (chart.index !== H.hoverChartIndex) {
             this.onContainerMouseLeave({ relatedTarget: true } as any);
         }
         H.hoverChartIndex = chart.index;
+         */
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: Remove the block instead of commenting.

File: ts/parts/Pointer.ts (hunk 3 above)
Comment (bre1470) [reply ts/parts/Pointer.ts]: The event order between Firefox and Chrome after click scrolling is different. Firefox does not fire reliable a move event after releasing the click wheel. Chrome fires always down and click.

File: ts/parts/Pointer.ts (hunk 4 above)
Comment (bre1470) [reply ts/parts/Pointer.ts]: Just to make sure, I understand the issue, the original condition was already flawed, correct? So we need additionally some IE specific code here?

File: ts/parts/Pointer.ts (hunk 7 above)
Comment (bre1470) [reply ts/parts/Pointer.ts]: That is a type thing, because I liked to remove the `any` casts. Without normalize `onContainerMouseDown` has to accept regular `MouseEvent`s without normalizing. It is the same with the other pointer callbacks.

File: ts/parts/Interaction.ts (hunk 1 above)
Comment (bre1470) [reply ts/parts/Interaction.ts]: This comes from an old merge during the first try of solving the issue.

File: ts/parts/Pointer.ts (hunk 2 above)
Comment (bre1470) [reply ts/parts/Pointer.ts]: Same here.

**Comment by bre1470** [comment]
Please review again.

File: ts/parts/Pointer.ts [hunk 14]
```diff
@@ -623,18 +623,16 @@ class Pointer {
      * @private
      * @function Highcharts.Pointer#drop
      *
-     * @param {global.Event} e
-     *
-     * @return {void}
+     * @param {global.Event} evt
      */
-    public drop(e: Event): void {
+    public drop(evt: Event): void {
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: Why was this renamed to `evt`? In all other aspects, we use `e` for event arguments, so we should keep consistent whenever we can.

File: ts/parts/Pointer.ts [hunk 15]
```diff
@@ -1312,18 +1306,19 @@ class Pointer {
...
      *
-     * @param {Highcharts.PointerEventObject} e
+     * @param {global.MouseEvent} evt
      *
      * @return {void}
      */
-    public onContainerMouseLeave(e: Highcharts.PointerEventObject): void {
+    public onContainerMouseLeave(evt: MouseEvent): void {
         const chart = charts[pick(H.hoverChartIndex, -1)];
+        const pEvt = this.normalize(evt);
```
Comment (TorsteinHonsi) [inline ts/parts/Pointer.ts]: Do we need to create a new variable for the normalized event arguments?
